main/lsp/requests/formatting/document_formatting.cc

```diff
@@ -0,0 +1,59 @@
+#include "main/lsp/requests/formatting/document_formatting.h"
+#include "common/common.h"
+#include "main/lsp/LSPOutput.h"
+#include "main/lsp/json_types.h"
+#include "main/lsp/lsp.h"
+
+using namespace std;
+
+namespace sorbet::realmain::lsp {
+DocumentFormattingTask::DocumentFormattingTask(const LSPConfiguration &config, MessageId id,
+                                               std::unique_ptr<DocumentFormattingParams> params)
+    : LSPRequestTask(config, move(id), LSPMethod::TextDocumentFormatting), params(move(params)) {}
+
+// Processed on the index thread so it doesn't wait for typechecking.
+LSPTask::Phase DocumentFormattingTask::finalPhase() const {
+    return Phase::INDEX;
+}
+
+void DocumentFormattingTask::index(LSPIndexer &index) {
+    auto response = make_unique<ResponseMessage>("2.0", id, LSPMethod::TextDocumentFormatting);
+    if (!config.opts.lspDocumentFormatRubyfmtEnabled) {
+        response->error = make_unique<ResponseError>(
+            (int)LSPErrorCodes::InvalidRequest,
+            "The `Document Formatting` LSP feature is experimental and disabled by default.");
+        config.output->write(move(response));
+        return;
+    }
+
+    variant<JSONNullObject, vector<unique_ptr<TextEdit>>> result = JSONNullObject();
+
+    auto fref = index.uri2FileRef(params->textDocument->uri);
+    if (fref.exists()) {
+        // Ensure that `rubyfmt` is in the user's path
+        auto rubyfmt_path = exec("which rubyfmt");
+        if (rubyfmt_path.empty()) {
```

Comment:
When you change this to use `subprocess`,  you should likely be able to catch thrown exceptions from the subprocess call to `rubyfmt` directly. This will let you delete this call to `which`. You can see in `WatchmanProcess.cc` how this works for detecting `watchman` (the example in that file uses `Popen` instead of what you probably want which is `check_output`, and as a result it has to deal with parent/child process IDs, but you should likely be able to simply catch results from check_output, ignoring anything about process IDs.)

Reply:
Took a crack at this, but unfortunately I think `check_output` only tells you if the process exits with a non-zero code, but it doesn't tell you _what_ that exit code is unless you use `Popen`, so I ended up going that route.